compiler/lsp-tests/src/Main.hs

```diff
@@ -387,25 +390,43 @@ stressTests
 stressTests run _runScenarios = testGroup "Stress tests"
   [ testCase "Modify a file 2000 times" $ run $ do
         let fooValue :: Int -> T.Text
+            -- Even values should produce empty diagnostics
+            -- while odd values will produce a type error.
             fooValue i = T.pack (show (i `div` 2))
-                      <> if i `mod` 2 == 0 then "" else ".5"
+                      <> if even i then "" else ".5"
             fooContent i = T.unlines
                 [ "daml 1.2"
                 , "module Foo where"
                 , "foo : Int"
                 , "foo = " <> fooValue i
                 ]
+            expect i = when (odd i) $ do
+                -- We do not wait for empty diagnostics on even i since debouncing
+                -- causes them to only be emitted after a delay which slows down
+                -- the tests.
+                -- Depending on debouncing we might first get a message with an empty
+                -- set of diagnostics or we might get the error diagnostics directly
+                -- if debouncing supresses the empty set of diagnostics.
+                diags <-
+                    skipManyTill
+                        anyMessage
+                        (do m <- LSP.message :: Session PublishDiagnosticsNotification;
+                            let diags = toList $ m ^. params . diagnostics
+                            guard (notNull diags)
+                            pure diags)
+                liftIO $ unless (length diags == 1) $
+                    assertFailure $ "Incorrect number of diagnostics, expected 1 but got " <> show diags
+                let msg = head diags ^. message
+                liftIO $ assertBool ("Expected type error but got " <> T.unpack msg) $
+                    "Couldn't match expected type" `T.isInfixOf` msg
         foo <- openDoc' "Foo.daml" damlId $ fooContent 0
-        forM_ [1 .. 999] $ \i ->
+        forM_ [1 .. 999] $ \i -> do
             replaceDoc foo $ fooContent i
-        -- We delay to account for debouncing
-        liftIO $ threadDelay (10 ^ (6 :: Int))
-        expectDiagnostics [("Foo.daml", [(DsError, (3, 6), "Couldn't match expected type")])]
-        forM_ [1000 .. 2000] $ \i ->
+            expect i
+        forM_ [1000 .. 2000] $ \i -> do
             replaceDoc foo $ fooContent i
-        liftIO $ threadDelay (10 ^ (6 :: Int))
+            expect i
```

Comment:
Why do we split this into two loops with the same body?

Reply:
Oops, that made sense before now it doesn’t, changed